Conditional event system

Frickerson opened the pull request.

Commit:
Create conditional event classes

Commit:
Implement activation and deactivation of different classes

Commit:
Replace references with pointers

Commit:
Fix macro errors

Commit:
Convert to subclass of

Commit:
Fix macros

Make sure to inherit from gas ext ability.
Add macro specifiers.

Commit:
Add getter for trigger manager comp

Commit:
Create simple trigger for trigger boxes

Commit:
Make subsystem functions blueprintcallable

Make sure game state is initialised before trying to access it.

Commit:
Make trigger classes not blueprintable

Commit:
Add Category for BPs

TheEmidee commented on Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalEvent.h:
```diff
@@ -0,0 +1,42 @@
+class UGBFConditionalTrigger;
+
+UCLASS( HideDropdown )
+class GAMEBASEFRAMEWORK_API UGBFConditionalEvent : public UGASExtGameplayAbility
```
Suffix with `Ability`

Frickerson replied:
Done.

TheEmidee commented on Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalEvent.h:
```diff
@@ -0,0 +1,42 @@
+
+protected:
+    UFUNCTION( BlueprintImplementableEvent, Category = "Conditional Events" )
+    void ApplyOutcomes();
```
what about `ExecuteOutcomes` ?

Frickerson replied:
Indeed a bit better.

TheEmidee commented on Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalEvent.h:
```diff
@@ -0,0 +1,42 @@
+
+private:
+    UFUNCTION()
+    void OnTriggerTriggered();
```
`OnTriggersTriggered`

Frickerson replied:
Done.

TheEmidee commented on Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.h:
```diff
@@ -0,0 +1,47 @@
+class AGameStateBase;
+
+UCLASS()
+class GAMEBASEFRAMEWORK_API UGBFConditionalEventSubsystem : public UWorldSubsystem
```
Implement `DoesSupportWorldType` (check the game phase subsystem)

Frickerson replied:
Done

TheEmidee commented on Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalTrigger.h:
```diff
@@ -0,0 +1,47 @@
+
+DECLARE_DYNAMIC_MULTICAST_DELEGATE( FGBFOnTriggeredDelegate );
+
+UCLASS( NotBlueprintable, EditInlineNew, HideDropdown )
```
Abstract

Frickerson replied:
Done.

TheEmidee commented on Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalTrigger.h:
```diff
@@ -0,0 +1,47 @@
+    GENERATED_BODY()
+
+public:
+    virtual void Activate();
```
PURE_VIRTUAL (both)

Frickerson replied:
Done.

TheEmidee commented on Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalTrigger.h:
```diff
@@ -0,0 +1,47 @@
+}
+
+UCLASS()
+class GAMEBASEFRAMEWORK_API UGBFTriggerBoxTrigger : public UGBFConditionalTrigger
```
Create a `Triggers` subfolder and put it there (you can leave the base class here)

Frickerson replied:
Done.

TheEmidee commented on Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalTrigger.h:
```diff
@@ -0,0 +1,47 @@
+}
+
+UCLASS()
+class GAMEBASEFRAMEWORK_API UGBFTriggerBoxTrigger : public UGBFConditionalTrigger
```
final

Frickerson replied:
Done.

TheEmidee commented on Source/GameBaseFramework/Public/Gameplay/GBFTriggerBox.h:
```diff
@@ -23,6 +23,8 @@ class GAMEBASEFRAMEWORK_API AGBFTriggerBox : public AActor
     UFUNCTION( NetMulticast, reliable, BlueprintCallable )
     void Multicast_DeactivateTrigger();
 
+    UGBFTriggerManagerComponent * GetTriggerManagerComponent() const;
```
inline functions are declared first

Frickerson replied:
Done.

TheEmidee commented on Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp:
```diff
@@ -0,0 +1,137 @@
+
+#include <AbilitySystemComponent.h>
+
+bool UGBFConditionalEventSubsystem::ShouldCreateSubsystem( UObject * outer ) const
```
On Lyra, the phase subsystem does not implement that function. The system would be created everywhere. But you need to add `BlueprintAuthorityOnly` on the BP functions

Frickerson replied:
Done.

TheEmidee commented on Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp:
```diff
@@ -0,0 +1,137 @@
+
+    if ( auto * world = GetWorld() )
+    {
+        world->GameStateSetEvent.AddUObject( this, &ThisClass::OnGameStateSet );
```
with game features being enabled / disabled at runtime, you can't cache the game state ASC like that.
You will need to get it like that everytime you need it : `auto * game_state_asc = world->GetGameState()->FindComponentByClass< UGASExtAbilitySystemComponent >();`

Frickerson replied:
Done.

TheEmidee commented on Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp:
```diff
@@ -0,0 +1,137 @@
+
+    if ( const auto * spec = GameStateASC->FindAbilitySpecFromClass( conditional_event ) )
+    {
+        GameStateASC->CancelAbilityHandle( spec->Handle );
```
you need to remove the Handle from the array

Frickerson replied:
Done.

TheEmidee commented on Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp:
```diff
@@ -0,0 +1,137 @@
+
+    for ( auto & event_handle : ActivatedEventHandles )
+    {
+        GameStateASC->CancelAbilityHandle( event_handle );
```
clear the array of handles

Frickerson replied:
Done.

TheEmidee commented on Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp:
```diff
@@ -0,0 +1,137 @@
+    }
+}
+
+void UGBFConditionalEventSubsystem::OnGameStateSet( AGameStateBase * game_state_base )
```
should be removed

Frickerson replied:
Done.

TheEmidee commented on Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalTrigger.cpp:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    auto * trigger_box = TriggerBoxSoftObject.Get();
+    trigger_box->Multicast_ActivateTrigger( true );
```
I disagree with this. We will never find out why a disabled trigger suddenly gets activated. If the trigger is not active, so be it, the trigger won't execute

Frickerson replied:
Done.

TheEmidee commented on Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalTrigger.cpp:
```diff
@@ -0,0 +1,40 @@
+
+    auto * trigger_box = TriggerBoxSoftObject.Get();
+    trigger_box->GetTriggerManagerComponent()->OnTriggerBoxActivated().RemoveDynamic( this, &ThisClass::OnTriggerBoxActivated );
+    trigger_box->Multicast_DeactivateTrigger();
```
same

Frickerson replied:
Done.

Commit:
PR fixes

Rename conditional event to add suffix ability.
Rename apply outcomes to execute outcomes.
Rename trigger triggered to triggers triggered.
Implement does support world type.
Add pure virtual to activate and deactivate.
Add final to trigger.
Move inline function to top.

Commit:
PR fixes

Remove ShouldCreateSubsystem.
Remove Initialize and OnGameStateSet.
Create function to get game state asc.
Remove handles when canceling abilities.
Add BlueprintAuthorityOnly to function calls.

Commit:
Do not activate or deactive trigger boxes

Commit:
Move triggerbox trigger out of trigger base class

Create own file in new folder for the class.

Commit:
Fixup const issues.

Commit:
Merge remote-tracking branch 'origin/develop' into conditional_event_system

Commit:
Fix trigger checks

Commit:
Add data validation

Commit:
Add non empty array checks

Commit:
Remove trigger validation

Validation does not work with inline classes.

TheEmidee commented on Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp:
```diff
@@ -0,0 +1,124 @@
+
+void UGBFConditionalEventSubsystem::ActivateEventGroup( const UGBFConditionalEventGroupData * conditional_event_group_data )
+{
+    const auto * game_state_asc = GetGameStateASC();
```
you don't need lines from L36 to L42

Frickerson replied:
Done.

TheEmidee commented on Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp:
```diff
@@ -0,0 +1,124 @@
+
+void UGBFConditionalEventSubsystem::DeactivateEventGroup( const UGBFConditionalEventGroupData * conditional_event_group_data )
+{
+    const auto * game_state_asc = GetGameStateASC();
```
don't need ASC here

Frickerson replied:
Done.

Commit:
PR fixes

remove unnecessary code
